Ingo 260512 (#54)
IngJohEricsson/draft-johansson-ccwg-scream-bis: draft-ietf-ccwg-rfc8298bis-screamv2.md

controller limits the sending rate so
that the number of bytes in flight is less than the reference window albeit with
a slack to avoid that packets are unnecessarily delayed in the Data Units Queue.

> The slack avoids unnecessary delays in the Data Units Queue when the link is uncongested. The bytes in flight however, become more restrained in congested situations, to avoid large variations in queue delay and bitrate.

Reviewer: ```suggestion
The slack avoids unnecessary delays in the Data Units Queue when the link is uncongested. The bytes in flight, however, become more restrained in congested situations to avoid large variations in queue delay and bitrate.
```

the difference in time between the event that the data unit was marked as lost and
the event that it was indicated as successfully received. Loss is detected if a
given data unit is not acknowledged within a time window (indicated by the
reordering window) after a data unit with a higher sequence number was
acknowledged.

A loss rate is calculated for each packet according to the equation below, the loss rate is calculated as an average with a time constant of 2 RTTs.

>  For small ref_wnd, averaging is over 4/LOSS_RATE_THRESHOLD packets, to avoid that consecutive link layer losses cause the loss rate to increase too quickly.

Reviewer: ```suggestion
A loss rate is calculated for each packet with an average value reflecting the number of packets send with in two  RTTs or 4/LOSS_RATE_THRESHOLD packets for small ref_win values. This avoids that multiple consecutive link layer losses cause the loss rate to increase too quickly.
```

 The loss_rate is used in {{link-loss-rate-policer}}. The averaging is selected to be slow enough to avoid that single packet drops cause backoff, provided that the queue delay is low, and fast enough to avoid excessive packet loss in the presence of very shallow queues.

~~~

# Apply an EWMA filter with a 2 RTT time constant
# or at least 4/LOSS_RATE_THRESHOLD packets.

> alpha = min(LOSS_RATE_THRESHOLD / 4, mss / ref_wnd / 2)

Reviewer: I not sure I understand this: shouldn't alpha be a value between 0 and 1? But mss / ref_wnd / 2 could be larger than one, no?

Also with is LOSS_RATE_THRESHOLD devided by 4? That seems like a very random value...

Also as the loss rate is not used to calculate the ref window in section #ref-wnd-update, this whole part should be moved into #link-loss-rate-policer. let's try to keep it as straight forward and simple as possible here.

Author: The intention with the /4 is to make the averaging slow enough when the reference window is small and avoid that link layer losses at very low bitrates trigger congestion backoff
mss/ref_wnd/2 should be less than 1 as ref_wnd can't be smaller than mss

Yes, I'll try and put it in #link-loss-rate-policer instead

Author: Fixed

The variable qdelay_dev_avg indicates the delay jitter, or to be more concrete, a moving average over the difference between qdelay_max_avg and qdelay_min_avg. Based on qdelay_dev_avg a scaling factor ref_wnd_delay_scale is calculated that is then applied to the reference window increase and the reference window headroom. The ref_wnd_delay_scale is 1.0 when the jitter is zero and decreases to 0.0 when the jitter grows to QDELAY_DEV_THRESHOLD or above. The scaling factor decreases when the delay jitter increases to limit the congestion reaction to short term delay variations and slow down the reference window growth for better stability.

~~~
function calculate_ref_wnd_delay_scale()
  # Reduce average max queue delay and move min average
  # queue delay towards the max
  qdelay_max_avg = qdelay_max_avg * (1.0 - QDELAY_MIN_MAX_AVG_G)

>   qdelay_min_avg = qdelay_min_avg * (1.0 - QDELAY_MIN_MAX_AVG_G) +

Reviewer: Why do you apply these steps? What's the reason for this?

It's a bit confusing that you overwrite qdelay_max_avg/qdelay_min_avg here. Maybe it would better to use a new variable like qdelay_max_avg_scaled/qdelay_min_avg_scaled?

Author: It is not really an overwrite. 
The max value will gradually decrease down to zero, but will be updated to a higher value if new incoming feedback indicates a higher value.
The min value will gradually increase up to the max value, but can be updated to a lower value if new incoming feedback has a lower qdelay.
In a perfect link with zero congestion jitter and zero clock drift, both the min and max values will be zero

Reviewer: But you also set qdelay_max_avg/qdelay_min_avg in line 622 and 621. that doesn't seem right to me then...?

Author: These update to the min and max values, the code on lines 674 and 675 does the kind of forgetting about the min and max

~~~
# Update min and max average queue delay for every ACKed RTP packet
qdelay_max_avg = min(qdelay_target, max(qdelay, qdelay_max_avg)))
qdelay_min_avg = min(qdelay, qdelay_min_avg))

if (now - last_update_qdelay_avg_time >= min(virtual_rtt, s_rtt))
  # Calculate qdelay_avg
  if (qdelay < qdelay_avg)
    qdelay_avg = qdelay
  else
    qdelay_avg = QDELAY_AVG_G * qdelay +
      (1.0 - QDELAY_AVG_G) * qdelay_avg
  end

  # Optional code to calculate the variation on queue delay, which is an
  # Indication of congestion or near congestion.
  if (REDUCE_JITTER == true)
    # Reduce average max queue delay and move min average
    # queue delay towards the max
    qdelay_max_avg = qdelay_max_avg * (1.0 - QDELAY_MIN_MAX_AVG_G)
    qdelay_min_avg = qdelay_min_avg * (1.0 - QDELAY_MIN_MAX_AVG_G) +
      qdelay_max_avg * QDELAY_MIN_MAX_AVG_G

>     calculate_ref_wnd_delay_scale()

Reviewer: Why is this added here and not below?

Author: Good question, will fix this

Author: Fixed

{{RFC6817}}. A qdelay sample is obtained for each received acknowledgement.
It is typically sufficient with one update per received acknowledgement.

* qdelay_max_avg (qdelay_target): Max queue delay

* qdelay_min_avg (0.0): Min queue delay

* last_update_qdelay_avg_time (0.0): Last time qdelay_avg was updated [s]

* s_rtt (0.0): Smoothed RTT [s], computed with a similar method to that
  described in {{RFC6298}}

The following constants are used:

* QDELAY_AVG_G (1.0/4): Exponentially Weighted Moving Average (EWMA) factor for qdelay_avg

* REDUCE_JITTER (false): (optional) config knob to enable jitter filtering

* QDELAY_MIN_MAX_AVG_G (1.0/16): Filter gain for qdelay_min_avg and qdelay_max_avg

The SCReAM algorithm can be further improved for a greater rate stability by taking variations in qdelay into consideration. The goal is to react less to delay variations, caused by e.g. link layer related scheduling and retransmissions, but still be reactive to actual queue delay, caused by congestion. The code below provides a example implementation but more advanced statistical analysis can be considered.

> The variable qdelay_dev_avg indicates how much the queue delay varies. ref_wnd_delay_scale is a scale factor that is applied to the reference window increase and the reference window headroom, ref_wnd_delay_scale is 1.0 when the delay jitter is low decreases as the delay jitter increases.

Reviewer: ```suggestion
The variable qdelay_dev_avg indicates the delay jitter, or more concrete a moving average over the difference between qdelay_max_avg and qdelay_min_avghow. Based on qdelay_dev_avg a scaling factor ref_wnd_delay_scale is calculated that is then applied to the reference window increase and the reference window headroom. When the jitter is a below a threshold (QDELAY_DEV_THRESHOLD), no scaling is applied (ref_wnd_delay_scale is 1.0). The scaling factor decreases when the delay jitter increases to limit the congestion reaction to short term delay variations.
```

~~~
function calculate_ref_wnd_delay_scale()
  # Calculate ref_wnd_delay_scale, range [0.0 1.0]
  qdelay_dev_avg = (1.0 - QDELAY_DEV_AVG_G) * qdelay_dev_avg +
    QDELAY_DEV_AVG_G * (qdelay_max_avg - qdelay_min_avg)
  ref_wnd_delay_scale = max(0.0, min(1.0, 1.0 - qdelay_dev_avg / QDELAY_DEV_THRESHOLD))
end
~~~

The following variables are used:

* qdelay_dev_avg (0.0): indicates how much the queue delay varies[s]

* ref_wnd_delay_scale (1.0): A scale factor this applied to the ref_wnd increase as well as the reference window headroom

The following constants are used:

* QDELAY_DEV_AVG_G (1.0/32): Exponentially Weighted Moving Average (EWMA) factor for qdelay_dev_avg

> * QDELAY_DEV_THRESHOLD (0.01): Threshold for qdelay_dev_avg [s]

Reviewer: Why is this threshold 0.01?

Author: 10ms is to give a reasonable robustness against scheduling jitter. It looks to work OK in the 5G simulator

### Reference Window Update {#ref-wnd-update}

The reference window update contains two parts. One that reduces the reference
window when congestion events (listed above) occur, and one part that
continuously increases the reference window.

> The reference window reduction takes effect whenever a congestion event is detected. A congestion event is either packet loss, ECN-CE marked packets or increased queue delay. Reference window reduction is limited to no more often than once every min(VIRTUAL_RTT, s_rtt) to allow for the reference window reduction to take effect.

Reviewer: ```suggestion
The reference window is reduced whenever a congestion event is detected. A congestion event is either packet loss, ECN-CE marked packets or increased queue delay. The reference window reduction is applied at most once every (smoothed RTT) or, to improve fairness to other flows with a different RTT, at least very VIRTUAL_RTT second.
```

> The reference window increase takes place every s_rtt. The reference window can thus increase even when congestion events occur. This makes it possible for SCReAM to achieve the L4S scalable congestion control property where one MSS increase of the reference window per RTT is offset by to marked packets per RTT.

Reviewer: ```suggestion
The reference window is increased once per smother RTT (s_rtt). The reference window can also increase even when a congestion event occurs. Their provided the L4S scalable congestion control property where a one-MSS-increase of the reference window per RTT is offset by two marked packets per RTT.
```

end

if (is_loss_t || is_ce_t || is_virtual_ce_t)

>   if (now - last_ref_wnd_i_update_time > 10 * s_rtt)

Reviewer: Should 10 be a (configurable) constant? Also why 10?

Author: 10 is something based on experiments, WebRTC-SCReAM has a different approach suggested by Per Kjellander that is actually more simple. It could be good to replace with that alternative, but I need to try it out first.

<img width="756" height="242" alt="image" src="https://github.com/user-attachments/assets/8813d5bd-59c4-48d6-bcba-93b7d169635c" />

Author: I actually tried this in the 5G simulator and it appears to work fine and is more simple. But it can be good to keep the existing code in the spec yet some time as I want to try it out some more.

    # L4S mode
    backoff_t = l4s_alpha / 2

    # Reduce backoff when target bitrate is lower than the ACKnowledged rate
    # this is indicative of that the target rate is already reduced

>     # by increased RTT

Reviewer: ```suggestion
    # this may be an indication that the target rate is already reduced
    # due to an increased RTT
```

    if (target_bitrate < acked_bitrate * ACKED_BITRATE_MARGIN)
        backoff_t *= BACKOFF_SCALE_LOW_TARGET_RATE
    end

    # Scale down backoff when RTT is high to avoid overreaction to
    # congestion. This is related to that congestion backoff can occur
    # every min(VIRTUAL_RTT, s_rtt)

>     backoff_t /= max(1.0, s_rtt / VIRTUAL_RTT)

Reviewer: Isn't this to implement Prague's fairness scaling?

Author: Yes, that is true. 
Also, for the case that RTT is large, then backoff happens more than once per RTT, therefore the backoff needs to be reduced.

if (is_virtual_ce_t)
  backoff_t = l4s_alpha_v_t / 2

  # Reduce backoff when target bitrate is lower than the ACKnowledged rate
  # this is indicative of that the target rate is already reduced
  # by increased RTT
  if (target_bitrate < acked_bitrate * ACKED_BITRATE_MARGIN)
     backoff_t *= BACKOFF_SCALE_LOW_TARGET_RATE

>   end

Reviewer: This part is here twice; see line 839

end
~~~

> The reference window reduction when congestion due to L4S marking or increased delay is detected is constrained by a number of factors:

Reviewer: ```suggestion
The reference window reduction, when congestion is detected due to L4S marking or increased delay, is constrained by a number of factors:
```

> * For cases that the target rate is lower than the acknowledged bitrate, it is likely that the target bitrate is already reduced because of an increased RTT. A further (large) decrease of the reference window would then only lead to an unnecessary undershoot.

Reviewer: Is this really a common case? This can only happen if the sender to receiver delay suddenly increases, no?

Author: Yes, the intention is to only reduce the backoff if an RTT increase (increased one way delay) has already pushed the target bitrate down.

Reviewer: I mean if this is a corner case, do we really need to add it to the base algorithm or more as an optional feature?

Author: It is not really a corner case as increased RTT during congestion happens also with queue based L4S, for instance when link capacity drops quickly. I think however, that it is OK to leave this as optional in the specification.

Author: I moved this to a separate section as optional

  transmitted. Updated when ref_wnd is updated and when data unit is
  transmitted [byte].

> The sender transmission control is executed for each transmitted packet, the send interval is dictated by the pacing but is restricted by the send window. In practice it however can be more practical to transmit packets in small burst to save processing power, or alternatively, to offload the pacing to OS kernel function.

Reviewer: ```suggestion
The sender transmission control is executed for each transmitted packet, the send interval is dictated by the pacing but is restricted by the send window. In practice it, however, can be more practical to transmit packets in small burst to save processing power, or alternatively, to offload the pacing to OS kernel function.
```

## Media Rate Control {#media-rate-control-2}

The media rate control algorithm is executed whenever the reference window is

> updated and calculates the target bitrate. It is however not necessary to execute this part more often than the frame period of the media encoder.

Reviewer: ```suggestion
updated and calculates the target bitrate. It is, however, not necessary to execute this part more often than the frame period of the media encoder.
```
